pkg/platform/authentication/auth.go

```diff
@@ -215,6 +249,57 @@ func (s *Auth) AuthenticateWithToken(token string) error {
 	})
 }
 
+// AuthenticateWithDeviceCode posts a request to authenticate this device on the Platform and waits
+// for the user to authorize the request.
+// The given callback function is called when the user should be prompted to authorize the request.
+// Returns an error if authentication cannot be performed (e.g. timeout or Platform is unreachable).
+func (s *Auth) AuthenticateWithDevice(promptCallback func(userCode, uri string)) error {
+	if s.Authenticated() {
+		return nil // nothing to do
+	}
+	// Post the authentication request to the Platform.
+	postParams := oauth.NewAuthDevicePostParams()
+	response, err := mono.Get().Oauth.AuthDevicePost(postParams)
+	if err != nil {
+		logging.Error("Error requesting device authentication: %v", err)
+		return locale.NewError("err_auth_device")
+	}
+	// Prompt the user to authorize the request.
+	promptCallback(*response.Payload.UserCode, *response.Payload.VerificationURIComplete)
+	// Wait for the authorization.
+	deviceCode := strfmt.UUID(*response.Payload.DeviceCode)
+	getParams := oauth.NewAuthDeviceGetParams()
+	getParams.SetDeviceCode(deviceCode)
+	startTime := time.Now()
+	const timeout = 6 * 60 * time.Second
+	for {
+		response, err := mono.Get().Oauth.AuthDeviceGet(getParams)
+		if response != nil {
+			s.storeAuthenticatedDevice(deviceCode, response)
+			break
+		} else if errs.Matches(err, &oauth.AuthDeviceGetBadRequest{}) {
+			badRequest := err.(*oauth.AuthDeviceGetBadRequest)
+			errorString := *badRequest.Payload.Error
+			const expiredToken = "expired_token"
+			const invalidClient = "invalid_client"
+			const slowDown = "slow_down"
+			if errorString == expiredToken || time.Since(startTime) >= timeout {
```

Comment:
This isn't addressing http timeouts or anything else happening further down the chain.

Consider using a context with timeouts, eg. https://golang.cafe/blog/golang-context-with-timeout-example.html

Reply:
The underlying `mono.Oauth.AuthDevicePost()` function appears to have built-in timeout handling. I've turned off my computer's networking before, and in the middle of requests, and each time the state tool reports an error authenticating. I think it's working as expected.